bolt-simple-taproot.md

```diff
@@ -0,0 +1,943 @@
+
+We add `option_simple_taproot` to the set of defined channel types.
+
+#### `open_channel` Extensions
+
+1. `tlv_stream`: `open_channel_tlvs`
+2. types:
+    1. type: 2 (`local_musig2_pubnonce`)
+    2: data:
+        * [`66*byte`:`nonces`]
+    1. type: 4 (`remote_musig2_pubnonce`)
+    2: data:
+        * [`66*byte`:`nonces`]
+
+##### Requirements
+
+The sending node:
+
+  - MUST specify the `local_musig2_pubnonce` and `remote_musig2_pubnonce` TLV
+    fields.
+
+The sending node SHOULD:
+
+  - use the `NonceGen` algorithm defined in `bip-musig2` to generate
+    `lcaol_musig2_pubnonce` and `remote_musig2_pubnonce` to  ensure it
+    generates nonces in a safe manner.
+
+The receiving node MUST fail the channel if:
+
+  - the message doesn't include a `local_musig2_pubnonce` and
+    `remote_musig2_pubnonce` value.
+
+  - a specified public nonce cannot be parsed as two compressed secp256k1
+    points
+
+##### Rationale
+
+For the initial Taproot channel type (`option_simple_taproot`), musig2 nonces
+must be exchanged in the first round trip (open->, <-accept) to ensure that the
+initiator is able to generate a valid musig2 partial signature at the next
+step.
+
+The `remote_musig2_pubnonce` will be used to sign the _remote_ party's
+commitment transaction, while the `local_musig2_pubnonce` will be used to sign
+the local party's commitment transaction. 
+
+We require two nonces, as there're actually two messages being signed: the
+local commitment and the remote commitment. If only one nonce was used, when a
+party went to sign and broadcast their local commitment, they would re-use a
+nonce, as the nonce was already used to sign the remote party's commitment.
+
+#### `accept_channel` Extensions
+
+1. `tlv_stream`: `accept_channel_tlvs`
+2. types:
+    1. type: 2 (`local_musig2_pubnonce`)
+    2: data:
+        * [`66*byte`:`nonces`]
+    1. type: 4 (`remote_musig2_pubnonce`)
+    2: data:
+        * [`66*byte`:`nonces`]
+
+##### Requirements
+
+The sender:
+
+ - MUST set `local_musig2_pubnonce` and `remote_musig2_pubnonce` to the
+   `musig2` public nonce as specified by the `NonceGen` algorithm of
+   `bip-musig2`
+
+ - Aggregate the received local+remote public nonces (into two combined nonces)
+   using the `NonceAgg` algorithm of `bip-musig2`.
+
+The receiver:
+
+ - MUST reject the channel if `local_musig2_pubnonce` or
+   `remote_musig2_pubnonce` are not present.
+
+#### `funding_created` Extensions
+
+The sender:
+
+  - MUST derive the aggregated `musig2` public key using the exchanged `funding_pubkey`s 
+    - Before aggregating the two funding public keys, the keys MUST be sorted
+      using the `KeySort` routine of `bip-musig2`.
+
+  - MUST use the exchanged `local_musig2_pubnonce` and `remote_musig2_pubnonce`
+    received from the remote party with the `NonceAgg` algorithm to derive the
+    combined nonce
+
+  - the `signature` field MUST be a `musig2` partial signature using the
```

Comment:
just as a sanity check for myself, could you verify that this field is now only 32 bytes (in accordance with the `Sign` method's return type), and therefore the deserialization of the funding_created and funding_signed messages must depend on whether or not they're referring to Taproot channels?

Reply:
Actually, reading your last paragraph:

> Re the partial sig serialization (need to catch up spec to this), I've opted to re-use the existing 64 byte sig field. Rather than just encoding the 32 byte s value, we'll also encode the combined nonce (R1||R2, before the blinding factor is applied) to fully utilize the space. This is useful, since you need this nonce to actually verify partial signatures in the first place (need to spell this out more in the spec draft).

It seems that `R1` and `R2` are each 33 bytes. Given that `s` already takes up 32 bytes of the originally 64-byte-long-field, I'm not sure there's enough space to squeeze in an extra 66 bytes.

Reply:
So I need to update this still, my last comment argues that we should just use the full 64 bytes here and send the `r` value along side `s`. 

> It seems that R1 and R2 are each 33 bytes. Given that s already takes up 32 bytes of the originally 64-byte-long-field, I'm not sure there's enough space to squeeze in an extra 66 bytes.

Made an error here translating from brain to writing. What I propose we encode is just `r`, which is the x-coordinate of the final `R` point used for the partial signature generation (the value encoded in the challenge hash). This is a field element, so it's 32 bytes, `s` (bounded by the group order) is also 32 bytes, so everythign fits.

Reply:
> and therefore the deserialization of the funding_created and funding_signed messages must depend on whether or not they're referring to Taproot channels?

Yes that's one tradeoff with re-using the field vs either: adding another field, or another message all together (would propagate to the ones you mention and the state machine ones as well). On the wire they can both be 64 bytes, but the mapping back to some sort of signature struct/object differs:
  * Schnorr: we have `(r, s)` one is a field element (x coord of the "final" nonce), the other bounded by the group size. 
  * ECDSA: we have `(r, s)` both values being `mod n` (bounded by group size) 

If we go this route, in Go, I'd likely do something like:
```go
type Sig interface {
    Verify(pub *btcec.PublicKey) error
    Encode(w io.Writer) error
    Decode(r io.Reader) error 
}

type CommitSig[S Sig] struct {
     Sig S 
}
```

Then depending on the context, I'd do:
  * regular chans: `CommitSig[*ecdsa.Signature]`
  * taproot chans: `CommitSIg[*schnorr.Signature]` 

With the type bubbling up to other fields where needed.

Reply:
So one major use case of having the aggregate combined final nonce (I opened a [naming issue](https://github.com/jonasnick/bips/issues/69), the ambiguity is killing me) is that its polarity dictates the sign of the secret nonce summands [here](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#signing). With 64 bytes in the `signature` field, only 32 would remain for `r`, so it would have to be an x-only-pubkey, and lose this particular utility of the sanity check.

Further, considering that it would neither be a valid signature, nor a type-conformant partial signature, I do feel like it would be misappropriating the field somewhat. My personal preference, if at the very least for the purpose of clarity and legibility for future protocol readers and developers, would be to introduce TLV extensions for these values, which would also allow the inclusion of the full 66-byte aggregate R-value-serialization.

Reply:
> So I need to update this still, my last comment argues that we should just use the full 64 bytes here and send the `r` value along side `s`.

Why? the final `r` value that could be published to the blockchain (i mean the final combined aggregated nonce or whatever it is called) can be calculated by both parties using the exchanged nonces, one party would never use the nonce he got alongside a `commit_sig` message. i am in favor of just only using 32 (the first 32 or the last 32) bytes of the space available with the partial sig. the other 32 should be stuffed with zeros or at least not untouched data from the ram (like c `malloc` or js `Buffer.allocUnsafe` do)

Reply:
> Why? the final r value that could be published to the blockchain (i mean the final combined aggregated nonce or whatever it is called) can be calculated by both parties using the exchanged nonces, one party would never use the nonce he got alongside a commit_sig message

The idea is that we might as well just use the space for something useful. This is the value that's needed to combine the partial signatures into an actual signature. Sending it is just a sort of echo that lets both sides verify this value. 

Alternatively, we could send the nonce that's needed to verify the partial signature sent. This would sorta be more useful, since a party can take this value directly w/o needing to maintain some extra state to know which nonce to use.

Reply:
> would be to introduce TLV extensions for these values, which would also allow the inclusion of the full 66-byte aggregate-nonce-serialization.

Well in order to generate a sig in the first place, both sides already need this value (which I think is the argument being made above). So then we'd be sending extra unnecessary bytes when we can just send 64-bytes minimally with useful information. 

If we are sending a value, then the most useful value would likely be the same nonce one needs to verify the incoming partial signature.

Reply:
Yeah, I was basically saying that I'd prefer the extra bytes to overloading the (de)serialization mechanism depending on the channel type; disregarding encoding, however, how should a node handle an `R` value that deviates from its expectation?